### PR TITLE
Added toolbar and add model tool

### PR DESCRIPTION
**Pull request by mbrousse**

### CODE REVIEW

File: flatland_viz/include/flatland_viz/flatland_viz.h [hunk 1]
```diff
@@ -48,27 +48,58 @@
...
+class WidgetGeometryChangeDetector;
 }
 
 class FlatlandWindow;
 
+// class VisualizationFrame : public QMainWindow, public WindowManagerInterface
+// class FlatlandViz : public QMainWindow {
+
+// class FlatlandViz : public QMainWindow, public WindowManagerInterface
+//{
```
Comment (josephduchesne) [inline flatland_viz/include/flatland_viz/flatland_viz.h]: Please remove the commented out test/debugging code

File: flatland_viz/include/flatland_viz/flatland_window.h [hunk 2]
```diff
@@ -44,24 +44,77 @@
...
+#include "rviz/tool_manager.h"
+#include "rviz/tool_properties_panel.h"
+#include "rviz/view_manager.h"
+#include "rviz/views_panel.h"
+#include "rviz/visualization_frame.h"
+#include "rviz/visualization_manager.h"
+#include "rviz/widget_geometry_change_detector.h"
+#include "rviz/yaml_config_reader.h"
+#include "rviz/yaml_config_writer.h"
+
```
Comment (josephduchesne) [inline flatland_viz/include/flatland_viz/flatland_window.h]: I don't think we actually need all of the above includes

File: flatland_viz/include/flatland_viz/flatland_window.h [hunk 3]
```diff
@@ -44,24 +44,77 @@
...
 class FlatlandWindow : public QMainWindow {
   Q_OBJECT
  public:
   FlatlandWindow(QWidget* parent = 0);
+  // rviz::ToolManager* tool_manager_;
+  // here
+  // rviz::VisualizationFrame* vFrame_;
```
Comment (josephduchesne) [inline flatland_viz/include/flatland_viz/flatland_window.h]: more test/debug commented out code to clean up

File: flatland_viz/include/flatland_viz/flatland_window.h [hunk 4]
```diff
@@ -44,24 +44,77 @@
...
+  rviz::VisualizationManager* visualization_manager_;
+  rviz::RenderPanel* render_panel_;
+  // QToolBar* toolbar_;
```
Comment (josephduchesne) [inline flatland_viz/include/flatland_viz/flatland_window.h]: more test/debug commented out code to clean up

File: flatland_viz/src/flatland_viz.cpp [hunk 5]
```diff
@@ -72,6 +99,8 @@ FlatlandViz::FlatlandViz(FlatlandWindow* parent) : QWidget((QWidget*)parent) {
   // Set the top-level layout for this FlatlandViz widget.
   setLayout(main_layout);
 
+  //&FlatlandViz::onToolbarRemoveTool);
```
Comment (josephduchesne) [inline flatland_viz/src/flatland_viz.cpp]: more test/debug commented out code to clean up

File: flatland_viz/src/flatland_viz.cpp [hunk 6]
```diff
@@ -80,9 +109,32 @@ FlatlandViz::FlatlandViz(FlatlandWindow* parent) : QWidget((QWidget*)parent) {
...
+
+  tool_man->addTool("flatland_viz/SpawnModel");
+
   manager_->startUpdate();
 
+  // Q_EMIT toolAdded( tool );
+  // addTool( "rviz/Interact" );
+  // addTool( "rviz/Select" );
+  // addTool( "rviz/SetInitialPose" );
+  // addTool( "rviz/SetGoal" );
```
Comment (josephduchesne) [inline flatland_viz/src/flatland_viz.cpp]: more test/debug commented out code to clean up

File: flatland_viz/src/flatland_viz.cpp [hunk 7]
```diff
@@ -107,6 +159,220 @@ FlatlandViz::FlatlandViz(FlatlandWindow* parent) : QWidget((QWidget*)parent) {
...
 }
 
+// Destructor.
+FlatlandViz::~FlatlandViz() {
+  delete render_panel_;
+  delete manager_;
+
+  // for (int i = 0; i < custom_panels_.size(); i++) {
+  //   delete custom_panels_[i].dock;
+  // }
```
Comment (josephduchesne) [inline flatland_viz/src/flatland_viz.cpp]: more test/debug commented out code to clean up

File: flatland_viz/src/flatland_viz.cpp [hunk 8]
```diff
@@ -107,6 +159,220 @@ FlatlandViz::FlatlandViz(FlatlandWindow* parent) : QWidget((QWidget*)parent) {
...
+
+  // delete panel_factory_;
```
Comment (josephduchesne) [inline flatland_viz/src/flatland_viz.cpp]: more test/debug commented out code to clean up

File: flatland_viz/src/flatland_viz.cpp [hunk 9]
```diff
@@ -107,6 +159,220 @@ FlatlandViz::FlatlandViz(FlatlandWindow* parent) : QWidget((QWidget*)parent) {
...
+  connect(toolbar_actions_, &QActionGroup::triggered, this,
+          &FlatlandViz::onToolbarActionTriggered);
+
+  add_tool_action_ = new QAction("", toolbar_actions_);
+  add_tool_action_->setToolTip("Add a new tool");
+  add_tool_action_->setIcon(rviz::loadPixmap("package://rviz/icons/plus.png"));
+  toolbar_->addAction(add_tool_action_);
+
+  // connect(add_tool_action_, SIGNAL(triggered()), this,
+  //        SLOT(openNewToolDialog()));
```
Comment (josephduchesne) [inline flatland_viz/src/flatland_viz.cpp]: more test/debug commented out code to clean up

File: flatland_viz/src/flatland_viz.cpp [hunk 10]
```diff
@@ -107,6 +159,220 @@ FlatlandViz::FlatlandViz(FlatlandWindow* parent) : QWidget((QWidget*)parent) {
...
+  remove_tool_menu_ = new QMenu();
+  QToolButton* remove_tool_button = new QToolButton();
+  remove_tool_button->setMenu(remove_tool_menu_);
+  remove_tool_button->setPopupMode(QToolButton::InstantPopup);
+  remove_tool_button->setToolTip("Remove a tool from the toolbar");
+  remove_tool_button->setIcon(
+      rviz::loadPixmap("package://rviz/icons/minus.png"));
+  toolbar_->addWidget(remove_tool_button);
+  // connect(remove_tool_menu_, SIGNAL(triggered()), this,
+  //        SLOT(onToolbarRemoveTool(QAction*)));
```
Comment (josephduchesne) [inline flatland_viz/src/flatland_viz.cpp]: more test/debug commented out code to clean up

File: flatland_viz/src/flatland_viz.cpp [hunk 11]
```diff
@@ -107,6 +159,220 @@ FlatlandViz::FlatlandViz(FlatlandWindow* parent) : QWidget((QWidget*)parent) {
...
+  if (dialog->exec() == QDialog::Accepted) {
+    tool_man->addTool(class_id);
+  }
+  manager_->startUpdate();
+  activateWindow();  // Force keyboard focus back on main window.
+}
+
+void FlatlandViz::setHideButtonVisibility(bool visible) {
+  //   hide_left_dock_button_->setVisible(visible);
+  //   hide_right_dock_button_->setVisible(visible);
```
Comment (josephduchesne) [inline flatland_viz/src/flatland_viz.cpp]: I don't think we have any dock functionality

File: flatland_viz/src/flatland_viz_node.cpp [hunk 12]
```diff
@@ -77,6 +89,17 @@ int main(int argc, char** argv) {
   window = new FlatlandWindow();
   window->show();
```
Comment (josephduchesne) [inline flatland_viz/src/flatland_viz_node.cpp]: more test/debug commented out code to clean up (below)

File: flatland_viz/src/flatland_window.cpp [hunk 13]
```diff
@@ -45,31 +46,156 @@
...
+
+  // NewObjectDialog *dialog =
+  //     new NewObjectDialog(tool_man->getFactory(), "Tool", empty,
+  //                         tool_man->getToolClasses(), &class_id);
+  // manager_->stopUpdate();
+  // if (dialog->exec() == QDialog::Accepted) {
+  //   tool_man->addTool(class_id);
+  // }
+  // manager_->startUpdate();
+  // activateWindow();  // Force keyboard focus back on main window.
```
Comment (josephduchesne) [inline flatland_viz/src/flatland_window.cpp]: more test/debug commented out code to clean up (above)

File: flatland_viz/src/flatland_window.cpp [hunk 14]
```diff
@@ -45,31 +46,156 @@
...
+}
+
+rviz::VisualizationManager *FlatlandWindow::getManager() {
+  return visualization_manager_;
+}
+
+void FlatlandWindow::CreateModelDialog() {
```
Comment (josephduchesne) [inline flatland_viz/src/flatland_window.cpp]: This function doesn't seem super useful. Is it used or can it be deleted?

File: flatland_viz/src/flatland_window.cpp [hunk 15]
```diff
@@ -45,31 +46,156 @@
...
+// toolbar_->addWidget(remove_tool_button);
+// connect(remove_tool_menu_, SIGNAL(triggered(QAction *)), this,
+//         SLOT(onToolbarRemoveTool(QAction *)));
+//}
+
+FlatlandWindow::FlatlandWindow(QWidget *parent) : QMainWindow(parent) {
+  // vFrame_ = new rviz::VisualizationFrame();
+  // vFrame_->setApp(parent->getApp());
+  // render_panel_ = new rviz::RenderPanel();
+  // visualization_manager_ = new rviz::VisualizationManager(render_panel_);
```
Comment (josephduchesne) [inline flatland_viz/src/flatland_window.cpp]: more test/debug commented out code to clean up (above)

File: flatland_viz/src/flatland_window.cpp [hunk 16]
```diff
@@ -45,31 +46,156 @@
...
   fps_label_->setMinimumWidth(40);
   fps_label_->setAlignment(Qt::AlignRight);
   statusBar()->addPermanentWidget(fps_label_, 0);
 
-  // Set the title
+  // Set the main window properties
   setWindowTitle("Flatland Viz");
+  QFont font;
+  font.setPointSize(font.pointSizeF() * 0.9);
+
```
Comment (josephduchesne) [inline flatland_viz/src/flatland_window.cpp]: more test/debug commented out code to clean up (below)

File: flatland_viz/src/model_dialog.cpp [hunk 17]
```diff
@@ -0,0 +1,259 @@
...
+  srv.request.yaml_path = path_to_model_file.toStdString();
+  srv.request.pose.x = x_edit->text().toFloat();
+  srv.request.pose.y = y_edit->text().toFloat();
+  srv.request.pose.theta = a_edit->text().toFloat();
+
+  client = nh.serviceClient<flatland_msgs::SpawnModel>("spawn_model");
+
+  client.call(srv);
+}
+
```
Comment (josephduchesne) [inline flatland_viz/src/model_dialog.cpp]: more test/debug commented out code to clean up (below)

**Comment by mbrousse** [comment]
There seems to be two copies of the flatland logo.

I think number 2 is the one in use and number 1 can be deleted.

File: flatland_viz/include/flatland_viz/flatland_window.h (hunk 3 above)
Comment (mbrousse) [reply flatland_viz/include/flatland_viz/flatland_window.h]: remove here

File: flatland_viz/src/flatland_viz.cpp (hunk 6 above)
Comment (mbrousse) [reply flatland_viz/src/flatland_viz.cpp]: I think we should start with default tools for Rviz plus Spawn Model Tool.

**Pull request merged**